core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,538 @@
+	hash_pos: Vec<u64>,
+	hashes: Vec<Hash>,
+	leaf_pos: Vec<u64>,
+	leaf_data: Vec<T>,
+	proof: SegmentProof,
+}
+
+impl<T> Segment<T> {
+	/// Maximum number of leaves in a segment, given by `2**b`
+	fn segment_capacity(&self) -> u64 {
+		1 << self.identifier.height
+	}
+
+	/// Offset (in leaf idx) of first leaf in the segment
+	fn leaf_offset(&self) -> u64 {
+		self.identifier.idx * self.segment_capacity()
+	}
+
+	/// Number of leaves in this segment. Equal to capacity except for the final segment, which can be smaller
+	fn segment_unpruned_size(&self, last_pos: u64) -> u64 {
+		min(
+			self.segment_capacity(),
+			pmmr::n_leaves(last_pos).saturating_sub(self.leaf_offset()),
+		)
+	}
+
+	/// Whether the segment is full (size == capacity)
+	fn full_segment(&self, last_pos: u64) -> bool {
+		self.segment_unpruned_size(last_pos) == self.segment_capacity()
+	}
+
+	/// Inclusive range of MMR positions for this segment
+	fn segment_pos_range(&self, last_pos: u64) -> (u64, u64) {
+		let segment_size = self.segment_unpruned_size(last_pos);
+		let leaf_offset = self.leaf_offset();
+		let first = pmmr::insertion_to_pmmr_index(leaf_offset + 1);
+		let last = if self.full_segment(last_pos) {
+			pmmr::insertion_to_pmmr_index(leaf_offset + segment_size)
+				+ (self.identifier.height as u64)
+		} else {
+			last_pos
+		};
+
+		(first, last)
+	}
+
+	fn get_hash(&self, pos: u64) -> Result<Hash, SegmentError> {
+		self.hash_pos
+			.iter()
+			.zip(&self.hashes)
+			.find(|&(&p, _)| p == pos)
+			.map(|(_, &h)| h)
+			.ok_or_else(|| SegmentError::MissingHash(pos))
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: Readable + Writeable + Debug,
+{
+	fn empty(identifier: SegmentIdentifier) -> Self {
+		Segment {
+			identifier,
+			hash_pos: Vec::new(),
+			hashes: Vec::new(),
+			leaf_pos: Vec::new(),
+			leaf_data: Vec::new(),
+			proof: SegmentProof::empty(),
+		}
+	}
+
+	/// Generate a segment from a PMMR
+	pub fn from_pmmr<U, B>(
+		segment_id: SegmentIdentifier,
+		pmmr: &ReadonlyPMMR<'_, U, B>,
+		prunable: bool,
+	) -> Result<Self, SegmentError>
+	where
+		U: PMMRable<E = T>,
+		B: Backend<U>,
+	{
+		let mut segment = Segment::empty(segment_id);
+
+		let last_pos = pmmr.unpruned_size();
+		if segment.segment_unpruned_size(last_pos) == 0 {
+			return Err(SegmentError::NonExistent);
+		}
+
+		// Fill leaf data and hashes
+		let (segment_first_pos, segment_last_pos) = segment.segment_pos_range(last_pos);
+		for pos in segment_first_pos..=segment_last_pos {
+			if pmmr::is_leaf(pos) {
+				if let Some(data) = pmmr.get_data_from_file(pos) {
+					segment.leaf_data.push(data);
+					segment.leaf_pos.push(pos);
+					continue;
+				} else if !prunable {
+					return Err(SegmentError::MissingLeaf(pos));
+				}
+			}
+			// TODO: optimize, no need to send every intermediary hash
+			if prunable {
+				if let Some(hash) = pmmr.get_from_file(pos) {
+					segment.hashes.push(hash);
+					segment.hash_pos.push(pos);
+				}
+			}
+		}
+
+		if segment.leaf_data.is_empty() {
```

Comment:
What are the implications of not being able to request a fully pruned segment?
If the first 1024 outputs are spent and the segment size is 1024, how do we proceed?

Reply:
The first step in the new sync process will be to reconstruct the full bitmap (mmr) by downloading the segments. From that point on we know which segments will be empty, and just skip them.

Reply:
Yes of course, this makes sense if we have the bitmap (which itself is not pruned). :+1:
So in the pathological case of just a single unpruned leaf at the right hand side of the PMMR, this just results in a single segment being requested. We don't need the root of any of the fully pruned segments as this data is effectively included in the Merkle proof of that final segment.

Reply:
If the second segment is not fully spent, we download it, validate it and add the data to the PMMR we are building. This includes the root of the first segment since we obtained it from the merkle proof.